llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp

```diff
@@ -447,11 +448,9 @@ static bool canRewriteGEPAsOffset(Type *ElemTy, Value *Start, Value *Base,
         return false;
 
       if (auto *GEP = dyn_cast<GEPOperator>(V)) {
-        // We're limiting the GEP to having one index. This will preserve
-        // the original pointer type. We could handle more cases in the
-        // future.
-        if (GEP->getNumIndices() != 1 || !GEP->isInBounds() ||
-            GEP->getSourceElementType() != ElemTy)
+        // Only allow GEPs with at most one variable offset.
+        auto IsNonConst = [](Value *V) { return !isa<ConstantInt>(V); };
+        if (count_if(GEP->indices(), IsNonConst) > 1)
```

Comment:
Why did you drop the `!GEP->isInBounds()` check?

https://github.com/llvm/llvm-project/blob/b84cb9c2e672a39cb94f080b67c121ff5ae74879/llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp#L685-L694

> We've only looked through inbouds GEPs we know that we can't have overflow on either side.

Reply:
Thanks for catching this! I've added some additional inbounds tests in https://github.com/llvm/llvm-project/commit/57a0a9aadf1ec4c111e303b7f0fe76505b3ad137 so we have coverage for this case and have restored the check.